lib/ansible_tower_client/base_models/credential.rb

```diff
@@ -39,7 +39,9 @@ def credential_type
     end
 
     def vault_password
-      @data['vault_password'] ||= inputs.vault_password.to_s
+      @data.fetch('vault_password') do |key|
+        @data[key] = (inputs.vault_password if inputs.respond_to?(:vault_password)).to_s
```

Comment:
@AlexanderZagaynov can you just do `@data['vault_password'] ||= inputs.vault_password.to_s if inputs.respond_to?(:vault_password)` ?  I think that is a little more obvious.

Reply:
@agrare it won't work in desired way, it will do second part many times
```
irb(main):001:0> h = {}; h[:a] ||= b if false; h
=> {}
irb(main):002:0> h = {}; h[:a] ||= (b if false); h
=> {:a=>nil}
```

Reply:
@AlexanderZagaynov you mean it'll run `inputs.respond_to?(:valut_password)` even after `@data['vault_password']` is set?  That's probably okay but if you're worried about it you could do `@data['value_password'] = inputs.vault_password.to_s if !@data.key?('vault_password') && inputs.respond_to(:vault_password)`

Reply:
No, in your example this key will not be even set. 
Anyway, at the end of the day, you just trying to invent what I already have here. Why? :)

Reply:
Not sure what you mean it won't ever be set...

```
>> inputs = OpenStruct.new
>> inputs.vault_password = "abcd"
>> data = {}
>> data['value_password'] = inputs.vault_password.to_s if !data.key?('vault_password') && inputs.respond_to?(:vault_password)
=> "abcd"
>> data
=> {"value_password"=>"abcd"}
```

Reply:
```
irb(main):001:0> inputs = OpenStruct.new
=> #<OpenStruct>
irb(main):002:0> data = {}
=> {}
irb(main):003:0> data['vault_password'] = inputs.vault_password.to_s if !data.key?('vault_password') && inputs.respond_to?(:vault_password)
=> nil
irb(main):004:0> data
=> {}
```

Reply:
Oh you want it to be set to nil if it doesn't respond to it?

Reply:
like that, empty string to be more precise. lazy calculation with possible empty result.

Reply:
Do we have active_support here?  This sounds like its just `inputs.try(:vault_password).to_s`

Reply:
no, we have no active support here officially, and we should avoid it here, if I got that right: https://github.com/ansible/ansible_tower_client_ruby/pull/119#discussion_r213090751

Reply:
It sounds like it's important that `vault_password` is always a string, so you may want something more like this?
`@data["vault_password"] = (inputs.vault_password if inputs.respond_to?(:vault_password)).to_s unless @data["vault_password"].kind_of?(String)`

```ruby
irb(main):001:0> inputs = OpenStruct.new
=> #<OpenStruct>
irb(main):002:0> h = {}
=> {}
irb(main):003:0> h["vault_password"] = (inputs.vault_password if inputs.respond_to?(:vault_password)).to_s unless h["vault_password"].kind_of?(String)
=> ""
irb(main):004:0> h
=> {"vault_password"=>""}
irb(main):005:0> inputs.vault_password = "abc"
=> "abc"
irb(main):006:0> h["vault_password"] = (inputs.vault_password if inputs.respond_to?(:vault_password)).to_s unless h["vault_password"].kind_of?(String)
=> nil
irb(main):007:0> h
=> {"vault_password"=>""}


irb(main):001:0> inputs = OpenStruct.new
=> #<OpenStruct>
irb(main):002:0> inputs.vault_password = "abc"
=> "abc"
irb(main):003:0> h = {}
=> {}
irb(main):004:0> h["vault_password"] = (inputs.vault_password if inputs.respond_to?(:vault_password)).to_s unless h["vault_password"].kind_of?(String)
=> "abc"
irb(main):005:0> h
=> {"vault_password"=>"abc"}
```

Reply:
1) @bdunne it makes no sense to check type if we can simply check key's presence
2) all that logic that you and @agrare are trying to invent works already in a Ruby standard way with Hash#fetch https://devdocs.io/ruby~2.5/hash#method-i-fetch
I do not understand, what the problem to know how standard Ruby's features works?

Reply:
Why do you need to return "" instead of nil?

Using fetch to set a value if it doesn't exist is not a common pattern and takes extra cycles to figure out what you're doing where as `||=` is immediately obvious.  Its not wrong it just looks strange (just did a quick search and we don't do this anywhere in the codebase)

Reply:
@agrare yes, we need to have a string here, further code breaks otherwise.
If we didn't something before it doesn't means it has no right to be here. Think of it as of diversity :)

Reply:
>  yes, we need to have a string here, further code breaks otherwise.

Interesting, handling nil is something we should do in the future

> If we didn't something before it doesn't means it has no right to be here

But it is reason to ask why you chose to do it that way which is what we were doing above

Reply:
@agrare https://gist.github.com/AlexanderZagaynov/160d93b0395ca5fd6674d78b5d15987b

Reply:
and here: https://gist.github.com/stevenharman/5664318